Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -209,8 +210,60 @@ CFArrayRef _DWriteCopyFontNamesForFamilyName(CFStringRef familyName) {
-    return _DWriteGetFontPropertiesFromName(fontName)->familyName.get();
+CFStringRef _DWriteGetCompatibleFamilyName(CFStringRef fontName, IDWriteFontFace* fontFace) {
+    woc::unique_cf<CFStringRef> upperFontName(_CFStringCreateUppercaseCopy(fontName));
+
+    const auto& info = __GetSystemFontCollectionHelper()->GetFontPropertiesFromUppercaseFontName(upperFontName);
+    if (info) {
+        return info->familyName.get();
+    }
+
+    const auto& userFontInfo = __GetUserFontCollectionHelper()->GetFontPropertiesFromUppercaseFontName(upperFontName);
+    if (userFontInfo) {
+        if (!fontFace) {
+            return userFontInfo->familyName.get();
+        }
+
+        // Retrieve the font file that was used to create this font face
+        UINT32 fileCount;
+        fontFace->GetFiles(&fileCount, nullptr);
+        RETURN_RESULT_IF(fileCount == 0, userFontInfo->familyName.get());
+        if (fileCount > 1) {
+            CFIndex familyNameLength = CFStringGetLength(userFontInfo->familyName.get());
+            std::vector<UniChar> chars(familyNameLength + 1);
+            CFStringGetCharacters(userFontInfo->familyName.get(), { 0, familyNameLength }, chars.data());
+            TraceWarning(TAG, L"The font %ls has %d files, defaulting to the first", chars.data(), fileCount);
+            fileCount = 1;
+        }
+
+        ComPtr<IDWriteFontFile> fontFile;
+        RETURN_RESULT_IF_FAILED(fontFace->GetFiles(&fileCount, &fontFile), userFontInfo->familyName.get());
+
+        // Get the CGDataProvider used to register this font
+        // Note that GetReferenceKey will set fontProvider to a pointer to the file's internal CGDataProviderRef
+        CGDataProviderRef* fontProvider = nullptr;
+        UINT32 keySize;
+        RETURN_RESULT_IF_FAILED(fontFile->GetReferenceKey((const void**)&fontProvider, &keySize), userFontInfo->familyName.get());
+        RETURN_RESULT_IF(keySize != sizeof(fontProvider) || !fontProvider || !(*fontProvider), userFontInfo->familyName.get());
+
+        // From the CGDataProvider get the URL
+        CFURLRef fontURL = _CGDataProviderGetURL(*fontProvider);
+        RETURN_RESULT_IF(!fontURL, userFontInfo->familyName.get());
+
+        // Now construct the XAML compatible family name, which should be in the format
+        // ms-appx:///[PATH TO FONT FILE]/font_file.ttf#family_name
+        static const CFStringRef sc_prefix = CFSTR("ms-appx:///");
+        auto ret = woc::MakeAutoCF<CFMutableStringRef>(CFStringCreateMutableCopy(nullptr, 0, sc_prefix));
+        auto path = woc::MakeAutoCF<CFStringRef>(CFURLCopyFileSystemPath(fontURL, kCFURLWindowsPathStyle));
+        CFStringAppend(ret, path);
+
+        static const UniChar numberSign = '#';
+        CFStringAppendCharacters(ret, &numberSign, 1);
+        CFStringAppend(ret, userFontInfo->familyName.get());
+        return (CFStringRef)CFAutorelease(ret.detach());
+    }
+
+    return nullptr;
```

Comment:
curious why are you just mixing early returns and late returns.
be consistent. 

---
In reply to: [107775106](https://github.com/Microsoft/WinObjC/pull/2297#discussion_r107775106) [](ancestors = 107775106)